src/aero_kernel/src/syscall/ipc.rs

```diff
@@ -0,0 +1,161 @@
+    let payload =
+        validate_slice_mut(message as *mut u8, messagesiz).ok_or(AeroSyscallError::EINVAL)?;
+
+    let mut messagequeue = MESSAGES.lock();
+    match messagequeue.get_mut(&pid) {
+        Some(mq) => {
+            mq.queue.push_back(Message {
+                from: get_scheduler().current_task().pid().as_usize(),
+                data: payload.to_vec(),
+            });
+        }
+        None => {
+            messagequeue.insert(
+                pid,
+                MessageQueue {
+                    queue: [Message {
+                        from: get_scheduler().current_task().pid().as_usize(),
+                        data: payload.to_vec(),
+                    }]
+                    .into(),
+                },
+            );
+        }
+    }
+    // let bqueueitem = bqueue.get_mut(&pid).unwrap();
+    // bqueueitem.replace(Message {
+    //     from: get_scheduler().current_task().pid().as_usize(),
+    //     data: payload.to_vec(),
+    // });
+    BLOCK_QUEUE.notify_complete();
+    Ok(0)
+}
+pub fn recv(
+    pidptr: usize,
+    messageptr: usize,
+    messagemax: usize,
+    block: usize,
+) -> Result<usize, AeroSyscallError> {
+    let pid = get_scheduler().current_task().pid().as_usize();
+
+    if block == 0 {
+        // nonblocking read
+        let mut msgqueue = MESSAGES.lock();
+        match msgqueue.get_mut(&pid) {
+            Some(m) => {
+                let item = m
+                    .queue
+                    .pop_front()
+                    .expect("empty message queues should always be deleted!");
+                if item.data.len() > messagemax {
+                    m.queue.push_front(item);
+                    return Err(AeroSyscallError::E2BIG);
+                }
+                if m.queue.len() == 0 {
+                    msgqueue
+                        .remove(&pid)
+                        .expect("safety violation: modification of a value behind a mutex!");
+                }
+                return messagequeue_do_recieve(pidptr, messageptr, messagemax, item);
+            }
+            None => {
+                // nope
+                return Err(AeroSyscallError::EAGAIN);
+            }
+        }
+    }
+
+    let mut queue_map = BLOCK_QUEUE
+        .block_on(&MESSAGES, |msg| {
+            let mp = msg.get_mut(&pid);
+            match mp {
+                Some(mq) => match mq.queue.front() {
+                    Some(_) => true,
+                    None => false,
+                },
+                _ => false,
+            }
+        })
+        .unwrap();
+    
+    let mq = queue_map
+        .get_mut(&pid)
+        .expect("someone else stole our messagequeue!");
+    let msg = mq
+        .queue
+        .pop_front()
+        .expect("someone else stole our message!");
+    if msg.data.len() > messagemax {
+        mq.queue.push_front(msg);
+        return Err(AeroSyscallError::E2BIG);
```

Comment:
Here instead of a using the return keyword you could add an else block like so, I think it would look cleaner and more rust-y :^)
```rs
if msg.data.len() > messagemax {
    mq.queue.push_front(msg);
    Err(AeroSyscallError::E2BIG)
} else {
    if mq.queue.len() == 0 {
        queue_map
            .remove(&pid)
            .expect("safety violation: modification of a value behind a mutex!");
    }
    messagequeue_do_recieve(pidptr, messageptr, messagemax, msg)
}
```

Reply:
sure!